### PR TITLE
ARIA-92 Automatic operation task configuration

### PR DESCRIPTION
**Pull request by tliron**
Main changes:

1. Removed `runs_on` field from task model and API
2. Two new instantiations phase, one to find the host nodes, and the second to configure operation instances for the execution plugin
3. Parser fills in `configuration` field in `OperationTemplate` using `dependencies` in TOSCA syntax
4. Fixes to reqs-and-caps bugs

### CODE REVIEW

File: aria/modeling/orchestration.py [hunk 1]
```diff
@@ -200,15 +200,52 @@ class PluginBase(ModelMixin):
...
+    :ivar retry_count: How many retries occurred
+    :vartype retry_count: int
     """
 
     __tablename__ = 'task'
 
     __private_fields__ = ['node_fk',
                           'relationship_fk',
                           'plugin_fk',
-                          'execution_fk',
```
Comment (ran-z) [inline aria/modeling/orchestration.py]: ?

File: aria/modeling/orchestration.py [hunk 2]
```diff
@@ -258,29 +290,17 @@ def execution(cls):
...
-
-    due_at = Column(DateTime, nullable=False, index=True, default=datetime.utcnow())
-    started_at = Column(DateTime, default=None)
-    ended_at = Column(DateTime, default=None)
+    implementation = Column(String)
     max_attempts = Column(Integer, default=1)
-    retry_count = Column(Integer, default=0)
     retry_interval = Column(Float, default=0)
     ignore_failure = Column(Boolean, default=False)
+    due_at = Column(DateTime, nullable=False, index=True, default=datetime.utcnow())
```
Comment (ran-z) [inline aria/modeling/orchestration.py]: if you already do the separation to "state" fields and other, this should probably be under state.

File: aria/orchestrator/workflows/api/task.py [hunk 3]
```diff
@@ -65,86 +65,82 @@ def __init__(self,
...
         """
 
-        assert isinstance(actor, (models.Node, models.Relationship))
-        assert actor_type in ('node', 'relationship')
         assert interface_name and operation_name
-        assert runs_on in models.Task.RUNS_ON
         super(OperationTask, self).__init__()
 
         self.actor = actor
+        self.actor_type = actor_type
```
Comment (ran-z) [inline aria/orchestrator/workflows/api/task.py]: why are any changes needed here (besides runs_on)?

File: extensions/aria_extension_tosca/profiles/tosca-simple-1.0/interfaces.yaml [hunk 4]
```diff
@@ -63,24 +63,48 @@ interface_types:
     pre_configure_source:
       description: >-
         Operation to pre-configure the source endpoint.
+      _extensions:
+        default_dependencies:
+          - edge > source
```
Comment (ran-z) [inline extensions/aria_extension_tosca/profiles/tosca-simple-1.0/interfaces.yaml]: as we mentioned, these should not be a part of the dependencies. it should sit elsewhere.

Comment (ran-z) [inline extensions/aria_extension_tosca/profiles/tosca-simple-1.0/interfaces.yaml]: furthermore, for the time being, it's not "default", but rather the only place where one can determine where the operation shall run.

File: tests/modeling/test_models.py [hunk 5]
```diff
@@ -583,48 +583,49 @@ def test_node_model_creation(self, node_template_storage, is_valid, name, runtim
                    node_template_storage.service.list()[0]
 
 
-class TestNodeIP(object):
+class TestNodeHostAddress(object):
```
Comment (ran-z) [inline tests/modeling/test_models.py]: why change this everywhere?

File: tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml [hunk 6]
```diff
@@ -155,14 +155,34 @@ topology_template:
...
                       - max_instances: { greater_or_equal: 8 }
-    
+            relationship:
+              interfaces:
+                Configure:
+                  target_changed:
+                    implementation:
+                      primary: changed.sh
+                      dependencies:
+                        - edge > target
```
Comment (ran-z) [inline tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml]: remove

File: aria/parser/consumption/modeling.py [hunk 7]
```diff
@@ -145,6 +145,24 @@ def consume(self):
         self.context.modeling.instance.validate_capabilities()
 
 
+class FindHosts(Consumer):
```
Comment (ran-z) [inline aria/parser/consumption/modeling.py]: why is this implemented as another consumer?

File: aria/orchestrator/execution_plugin/instantiation.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ...utils.formatting import full_type_name
+from ...utils.collections import OrderedDict
+from ...parser import validation
+from ...parser.consumption import ConsumptionContext
+
+
+def configure_operation(operation):
```
Comment (ran-z) [inline aria/orchestrator/execution_plugin/instantiation.py]: almost everything in this module needs to go. this method is okay, the rest should be transparent to the execution plugin

File: aria/modeling/service_instance.py [hunk 9]
```diff
@@ -1407,6 +1441,19 @@ def operation_template_fk(cls):
 
     # endregion
 
+    def configure(self):
```
Comment (ran-z) [inline aria/modeling/service_instance.py]: this definitely cant be here.

File: aria/modeling/service_instance.py [hunk 10]
```diff
@@ -180,6 +181,16 @@ def validate_capabilities(self):
                 satisfied = False
         return satisfied
 
+    def find_hosts(self):
```
Comment (ran-z) [inline aria/modeling/service_instance.py]: more methods on the models which we'll want to remove in a short while :(
i mean, this method isn't relevant for an instantiated service - it's only used for its (and its dependencies) creation. why would it sit as part of the instance?

File: aria/modeling/service_instance.py [hunk 11]
```diff
@@ -529,6 +524,32 @@ def validate_capabilities(self):
...
+                        return host
+            for the_relationship in node.inbound_relationships:
+                if (the_relationship.target_capability is not None) and \
+                    the_relationship.target_capability.type.role == 'feature':
+                    host = _find_host(the_relationship.source_node)
+                    if host is not None:
+                        return host
+            return None
+
+        self.host = _find_host(self)
```
Comment (ran-z) [inline aria/modeling/service_instance.py]: why is this dynamic? are we expecting this to change for an existing node?

File: tests/modeling/test_models.py (hunk 5 above)
Comment (mxmrlv) [inline tests/modeling/test_models.py]: no new tests were written for the new feature.

File: aria/modeling/service_instance.py (hunk 10 above)
Comment (tliron) [inline aria/modeling/service_instance.py]: That's not true: this phase will need be run every time there is a change in the topology. This is part of the reason I don't think `instantiation` is the right name for this module we talk about, since it would actually be used in many cases after instantiation.

File: aria/modeling/orchestration.py (hunk 1 above)
Comment (tliron) [inline aria/modeling/orchestration.py]: The diff algorithm borked here, if you look at the file it makes sense.

File: aria/modeling/orchestration.py (hunk 2 above)
Comment (tliron) [inline aria/modeling/orchestration.py]: Does `due_at` ever change after creation? I guess you're implying that it does...

File: aria/modeling/service_instance.py (hunk 9 above)
Comment (tliron) [inline aria/modeling/service_instance.py]: OK, so where? We don't have these separate module yet and I did not think I should be creating it as part of this PR.

File: aria/orchestrator/execution_plugin/instantiation.py (hunk 8 above)
Comment (tliron) [inline aria/orchestrator/execution_plugin/instantiation.py]: OK, where do you suggest moving them?

File: aria/orchestrator/workflows/api/task.py (hunk 3 above)
Comment (tliron) [inline aria/orchestrator/workflows/api/task.py]: Because we decided together that the logic that was here should not be in operation creation, but in service instance creation.

File: aria/parser/consumption/modeling.py (hunk 7 above)
Comment (tliron) [inline aria/parser/consumption/modeling.py]: This is just how phases work right now in "the parser". We keep talking about all of this moving into a separate "instantiation" module, but this is what we have for now.

File: extensions/aria_extension_tosca/profiles/tosca-simple-1.0/interfaces.yaml (hunk 4 above)
Comment (tliron) [inline extensions/aria_extension_tosca/profiles/tosca-simple-1.0/interfaces.yaml]: I thought you said that if there is no cost to allowing the user to override it, we could.

File: tests/modeling/test_models.py (hunk 5 above)
Comment (tliron) [inline tests/modeling/test_models.py]: That whole part of the code (finding the "ip") had to be refactored because it was essentially broken. As I was changing it anyway it made sense to change the name. It's really not hard to change back.

File: aria/modeling/service_instance.py [hunk 12]
```diff
@@ -529,6 +524,32 @@ def validate_capabilities(self):
                 satisfied = False
         return satisfied
 
+    def find_host(self):
+        def _find_host(node):
```
Comment (tliron) [inline aria/modeling/service_instance.py]: If we do that then we'll be calling `find_host` again on again on the same nodes that are targets of different relationships. So, to do that efficiently we would need some extra flag (`found_host`?) to signify that we don't have to do that again for that node. It made more sense to me to have it set once and only once per node.

File: aria/modeling/service_instance.py [hunk 13]
```diff
@@ -529,6 +524,32 @@ def validate_capabilities(self):
...
+            if node.type.role == 'host':
+                return node
+            for the_relationship in node.outbound_relationships:
+                if (the_relationship.target_capability is not None) and \
+                    the_relationship.target_capability.type.role == 'host':
+                    host = _find_host(the_relationship.target_node)
+                    if host is not None:
+                        return host
+            for the_relationship in node.inbound_relationships:
+                if (the_relationship.target_capability is not None) and \
```
Comment (tliron) [inline aria/modeling/service_instance.py]: This is a bit hard to explain. We tend to think of the host relationship as "contained in", but TOSCA also allows for "hosts a" kind of relationship, for example for features. See [here](https://github.com/apache/incubator-ariatosca/blob/ARIA-92-plugin-in-implementation-string/tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml#L177): to find out the loadbalancer node's host, we need to look inbound that it is a feature of Nginx.

So, bottom line is that this function has to be bi-recursive, both inbound and outbound. Yay. :)

File: aria/modeling/service_instance.py (hunk 11 above)
Comment (tliron) [inline aria/modeling/service_instance.py]: I don't know what you mean by "dynamic" here. These functions must be called during the initial instantiation phase, and then will need to be called again whenever there is a change in the topology.

File: aria/modeling/service_template.py [hunk 14]
```diff
@@ -1544,11 +1547,24 @@ def as_raw(self):
...
         from . import models
+        from ..orchestrator import context
+        plugin = None
+        if self.plugin_specification is not None:
+            try:
+                workflow_context = context.workflow.current.get()
+                plugin = self.plugin_specification.find_plugin(workflow_context.model.plugin.list())
+            except context.exceptions.ContextException:
+                pass
+                # TODO
```
Comment (tliron) [inline aria/modeling/service_template.py]: Yes, we need to talk about this. :) I don't know how else to find other models. Is there a current context for SQLAlchemy?

File: aria/modeling/orchestration.py (hunk 2 above)
Comment (ran-z) [inline aria/modeling/orchestration.py]: yup :)

File: aria/modeling/service_instance.py [hunk 15]
```diff
@@ -522,17 +524,10 @@ def properties(cls):
...
-            return None
-        host_node = self.host
-        if 'ip' in host_node.runtime_properties:  # pylint: disable=no-member
-            return host_node.runtime_properties['ip']  # pylint: disable=no-member
-        host_node = host_node.node_template  # pylint: disable=no-member
-        host_ip_property = host_node.properties.get('ip')
-        if host_ip_property:
-            return host_ip_property.value
+    def host_address(self):
+        if self.host:
```
Comment (ran-z) [inline aria/modeling/service_instance.py]: &&? :P

File: aria/modeling/service_instance.py [hunk 16]
```diff
@@ -522,17 +524,10 @@ def properties(cls):
...
+            if self.host.runtime_properties:
+                return self.host.runtime_properties.get('host_address')
```
Comment (ran-z) [inline aria/modeling/service_instance.py]: while I understand your reasoning to using "host_address" instead of "ip" everywhere, the actual runtime property that gets set by plugins remains to be "ip". this can't change at this stage.

File: aria/modeling/service_instance.py (hunk 13 above)
Comment (ran-z) [inline aria/modeling/service_instance.py]: :+1:

File: aria/modeling/service_instance.py (hunk 12 above)
Comment (ran-z) [inline aria/modeling/service_instance.py]: im still a bit confused by this myself - caching the host is definitely a good idea, but it looks here like the internal method _find_host is called recursively repeatedly.
I'm not sure this is exactly the same as what Maxim meant, but I think I'd expect each node's host finding method to run once at most, so that would mean either "host" is a self-initializing property, or a field initialized at its constructor, or you could use the "find_host" method, but basically what i'm missing here is the check for whether the value has been cached already or not.

i'd expect it to be more like:
```
def find_host(self):
    def _find_host(node):
    ....
        host = the_relationship.target_node.find_host()
    ....

if not self.host:  # or indeed use some extra flag
    self.host = _find_host(node)    
return self.host
```

File: aria/modeling/service_instance.py [hunk 17]
```diff
@@ -1632,9 +1673,9 @@ def interface_fk(cls):
         return relationship.foreign_key('interface', nullable=True)
 
     @declared_attr
-    def plugin_specification_fk(cls):
-        """For Operation one-to-one to PluginSpecification"""
-        return relationship.foreign_key('plugin_specification', nullable=True)
+    def plugin_fk(cls):
```
Comment (ran-z) [inline aria/modeling/service_instance.py]: does it make sense for operation no longer to have an fk for plugin specification at all?
is that going to remain for operation template?

File: aria/modeling/service_instance.py [hunk 18]
```diff
@@ -1685,12 +1725,31 @@ def inputs(cls):
...
     implementation = Column(Text)
+    configuration = Column(modeling_types.StrictDict(key_cls=basestring))
     dependencies = Column(modeling_types.StrictList(item_cls=basestring))
     executor = Column(Text)
     max_retries = Column(Integer)
     retry_interval = Column(Integer)
 
+    def configure(self):
+        from . import models
+        if (self.implementation is None) or (self.interface is None):
```
Comment (ran-z) [inline aria/modeling/service_instance.py]: how can interface be None here?

File: aria/modeling/service_instance.py [hunk 19]
```diff
@@ -1685,12 +1725,31 @@ def inputs(cls):
...
+            return
+
+        if self.plugin is None:
+            arguments = configure_operation(self)
```
Comment (ran-z) [inline aria/modeling/service_instance.py]: please use `execution_plugin.instantiation.configure_operation` or so, there are enough `configure_operation` calls in this module as it is and it's easily confusing :)

File: tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml [hunk 20]
```diff
@@ -98,8 +98,8 @@ topology_template:
               #token: { get_property: [ HOST, flavor_name ] }
       interfaces:
         Maintenance:
-          enable: juju > charm.maintenance_on
-          disable: juju > charm.maintenance_off
+          enable: juju \> charm.maintenance_on
```
Comment (ran-z) [inline tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml]: wait so what's this symbol about again? :)
Why are we not using the same symbol as we do for `dependencies`?

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py [hunk 21]
```diff
@@ -78,6 +78,11 @@ def get_inherited_capability_definitions(context, presentation, for_presentation
                 #capability_definitions[capability_name] = capability_definition
             else:
                 capability_definition = our_capability_definition._clone(for_presentation)
+                if isinstance(capability_definition._raw, basestring):
+                    # # Make sure we have a dict
```
Comment (ran-z) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: double comment

File: aria/orchestrator/workflows/builtin/utils.py [hunk 22]
```diff
@@ -71,15 +71,13 @@ def relationship_tasks(
...
-                                           runs_on='source')
+                                           operation_name=source_operation_name)
         )
     if target_operation_name:
         operations.append(
             OperationTask.for_relationship(relationship=relationship,
                                            interface_name=interface_name,
-                                           operation_name=target_operation_name,
-                                           runs_on='target')
+                                           operation_name=target_operation_name,)
```
Comment (ran-z) [inline aria/orchestrator/workflows/builtin/utils.py]: tuple?

File: aria/parser/consumption/modeling.py (hunk 7 above)
Comment (ran-z) [inline aria/parser/consumption/modeling.py]: :+1:

File: tests/modeling/test_models.py (hunk 5 above)
Comment (ran-z) [inline tests/modeling/test_models.py]: :+1:

File: aria/modeling/orchestration.py (hunk 1 above)
Comment (ran-z) [inline aria/modeling/orchestration.py]: :+1:

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py [hunk 23]
```diff
@@ -352,20 +359,44 @@ def create_interface_template_model(context, service_template, interface):
...
+            model.plugin_specification = service_template.plugin_specifications.get(plugin_name)
+            if model.plugin_specification is None:
+                context.validation.report(
+                    'unknown plugin "%s" specified in operation implementation: %s'
+                    % (plugin_name, primary),
+                    locator=operation._get_child_locator('implementation', 'primary'),
+                    level=Issue.BETWEEN_TYPES)
+    
+        relationship_edge = operation._get_extensions(context).get('relationship_edge')
+        if relationship_edge is not None:
```
Comment (ran-z) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: what if it is None? where do we set the default? I think either we don't or I've missed it :)

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py [hunk 24]
```diff
@@ -352,20 +359,44 @@ def create_interface_template_model(context, service_template, interface):
...
-            parse_implementation_string(context, service_template, operation.implementation.primary)
-
+    if implementation is not None: 
+        primary = implementation.primary
+        plugin_name, model.implementation = split_prefix(primary)
+        if plugin_name is not None:
+            model.plugin_specification = service_template.plugin_specifications.get(plugin_name)
+            if model.plugin_specification is None:
+                context.validation.report(
+                    'unknown plugin "%s" specified in operation implementation: %s'
```
Comment (ran-z) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: is this message clear enough about the fact that the plugin is not properly declared as a policy in the service template (rather than the question of whether it exists or not in the plugin repository)?

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py [hunk 25]
```diff
@@ -461,8 +492,15 @@ def create_workflow_operation_template_model(context, service_template, policy):
     properties = policy._get_property_values(context)
     for prop_name, prop in properties.iteritems():
         if prop_name == 'implementation':
-            model.plugin_specification, model.implementation = \
-                parse_implementation_string(context, service_template, prop.value)
+            plugin_name, model.implementation = split_prefix(prop.value)
+            if plugin_name is not None:
```
Comment (ran-z) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: duplicate code, why not refactor this?

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py [hunk 26]
```diff
@@ -677,21 +715,18 @@ def pattern(node_type, container): # pylint: disable=unused-argument
...
-    if plugin_name == 'execution':
-        plugin_specification = None
-    else:
-        plugin_specification = service_template.plugin_specifications.get(plugin_name)
-        if plugin_specification is None:
-            raise ValueError('unknown plugin: "{0}"'.format(plugin_name))
 
-    implementation = implementation[index+1:].strip()
-    return plugin_specification, implementation
+def set_nested(the_dict, keys, value):
```
Comment (ran-z) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: could you add a few comments here, possibly an example of input --> output, just so it's clearer? thanks

File: aria/modeling/service_instance.py (hunk 10 above)
Comment (ran-z) [inline aria/modeling/service_instance.py]: :+1:

File: aria/modeling/service_instance.py (hunk 11 above)
Comment (ran-z) [inline aria/modeling/service_instance.py]: :+1:

File: aria/modeling/service_template.py [hunk 27]
```diff
@@ -1848,11 +1856,30 @@ def as_raw(self):
...
     def instantiate(self, container):
         from . import models
+        from ..orchestrator import context
+        plugin = None
+        if self.plugin_specification is not None:
+            # TODO: we are planning a separate "instantiation" module where this will be called or
+            # moved to. There, we will probably have a context with a storage manager. Until then,
+            # this is the only potentially available context, which of course will only be available
+            # if we're in a workflow.
+            plugin = None
```
Comment (ran-z) [inline aria/modeling/service_template.py]: redundant

File: aria/modeling/service_template.py [hunk 28]
```diff
@@ -1848,11 +1856,30 @@ def as_raw(self):
 
     def instantiate(self, container):
         from . import models
+        from ..orchestrator import context
+        plugin = None
+        if self.plugin_specification is not None:
```
Comment (ran-z) [inline aria/modeling/service_template.py]: I'm not sure I understand this section.
Operations are instantiated at service creation time. No workflow context should be available at this time.
If a storage model object is required, it should be passed into the instantiate module directly (the current interface is problematic as is, and I've already broken it on the CLI branch in order to pass inputs to service template's instantiate), not taken off some thread local context.

File: aria/orchestrator/workflows/api/task.py [hunk 29]
```diff
@@ -66,98 +66,84 @@ def __init__(self,
...
         assert interface_name and operation_name
-        assert runs_on in models.Task.RUNS_ON
         super(OperationTask, self).__init__()
 
+        operation = None
+        interface = actor.interfaces.get(interface_name)
+        if interface is not None:
+            operation = interface.operations.get(operation_name)
+
+        if operation is None:
```
Comment (ran-z) [inline aria/orchestrator/workflows/api/task.py]: (FYI this section is refactored elsewhere in the CLI branch, so it will get overridden here too)

File: aria/orchestrator/workflows/api/task.py [hunk 30]
```diff
@@ -66,98 +66,84 @@ def __init__(self,
...
+            raise exceptions.OperationNotFoundException(
+                'Could not find operation "{0}" on interface "{1}" for {2} "{3}"'
+                .format(operation_name, interface_name, actor_type, actor.name))
+
+        if operation.implementation is None:
```
Comment (ran-z) [inline aria/orchestrator/workflows/api/task.py]: Why should an empty operation raise an error (rather than do nothing)?
if a node template has its type as root, and implements `create` with a proper implementation but does not use `configure`, would it not appear as an empty implementation operation?

File: aria/modeling/service_instance.py (hunk 17 above)
Comment (tliron) [inline aria/modeling/service_instance.py]: Correct.

File: aria/modeling/service_instance.py (hunk 18 above)
Comment (tliron) [inline aria/modeling/service_instance.py]: It will be None if the operation is attached to `Service` as a workflow :) I know you suggested a different model entirely for workflows ... that would be a separate JIRA perhaps.

File: aria/modeling/service_instance.py (hunk 19 above)
Comment (tliron) [inline aria/modeling/service_instance.py]: :+1:

File: tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml (hunk 20 above)
Comment (tliron) [inline tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml]: I've put the backslash to escape it and disable it for now. :/ Since the Juju plugin is not installed, the tests that use this would fail otherwise.

File: aria/orchestrator/execution_plugin/instantiation.py [hunk 31]
```diff
@@ -0,0 +1,187 @@
...
+    from . import operations
+    operation.implementation = '{0}.{1}'.format(operations.__name__,
+                                                operations.run_script_locally.__name__)
+
+
+def _configure_remote(operation, configuration, arguments):
+    """
+    Remote SSH operation via Fabric.
+    """
+    default_user = 'admin'
```
Comment (ran-z) [inline aria/orchestrator/execution_plugin/instantiation.py]: why these defaults? wouldnt it be better to fail?

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py (hunk 21 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: :+1:

File: aria/orchestrator/execution_plugin/instantiation.py [hunk 32]
```diff
@@ -0,0 +1,187 @@
...
+    arguments['use_sudo'] = ssh.get('use_sudo')
+    arguments['hide_output'] = ssh.get('hide_output')
+    arguments['fabric_env'] = {}
+    if 'warn_only' in ssh:
+        arguments['fabric_env']['warn_only'] = ssh['warn_only']
+    arguments['fabric_env']['user'] = ssh.get('user')
+    arguments['fabric_env']['password'] = ssh.get('password')
+    arguments['fabric_env']['key'] = ssh.get('key')
+    arguments['fabric_env']['key_filename'] = ssh.get('key_filename')
+    if 'address' in ssh:
```
Comment (ran-z) [inline aria/orchestrator/execution_plugin/instantiation.py]: i assume you came up with this one?

File: aria/orchestrator/execution_plugin/instantiation.py [hunk 33]
```diff
@@ -0,0 +1,187 @@
...
+    Remote SSH operation via Fabric.
+    """
+    default_user = 'admin'
+    default_password = 'admin'
+    ssh = _get_ssh(configuration.pop('ssh')) if 'ssh' in configuration else {}
+    if 'user' not in ssh:
+        ssh['user'] = default_user
+    if ('password' not in ssh) and ('key' not in ssh) and ('key_filename' not in ssh):
+        ssh['password'] = default_password
+    arguments['use_sudo'] = ssh.get('use_sudo')
```
Comment (ran-z) [inline aria/orchestrator/execution_plugin/instantiation.py]: why not use copy of some sort?

File: aria/orchestrator/workflows/builtin/utils.py (hunk 22 above)
Comment (tliron) [inline aria/orchestrator/workflows/builtin/utils.py]: :+1: (though it's totally legally in Python)

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py (hunk 23 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: None is the default -- it means that nowhere did the user set the edge explicitly set one way or the other (the column is nullable).

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py (hunk 24 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: :+1:

File: aria/orchestrator/execution_plugin/instantiation.py [hunk 34]
```diff
@@ -0,0 +1,187 @@
...
+
+
+def _validate_type(value, the_type, name):
+    if not isinstance(value, the_type):
+        context = ConsumptionContext.get_thread_local()
+        context.validation.report('"{0}" configuration is not a {1}'
+                                  .format(name, full_type_name(the_type)),
+                                  level=validation.Issue.BETWEEN_TYPES)
+
+def _str_to_bool(value, name):
```
Comment (ran-z) [inline aria/orchestrator/execution_plugin/instantiation.py]: this and the next method need to sit elsewhere. on CLI branch there's a `type` module under `utils`, and the `dict_to_list`  should go to where there's already similar methods.

the validation thats called within these methods can be called before or after them

File: aria/orchestrator/execution_plugin/instantiation.py [hunk 35]
```diff
@@ -0,0 +1,187 @@
...
+                                      level=validation.Issue.BETWEEN_TYPES)
+    return value
+
+
+def _get_ssh(value):
+    if value is None:
+        return {}
+    _validate_type(value, dict, 'ssh')
+    for k, v in value.iteritems():
+        if k == 'use_sudo':
```
Comment (ran-z) [inline aria/orchestrator/execution_plugin/instantiation.py]: i think it'd probably have been neater to have a dict mapping field-name to type and then validating it rather than having to add here manually every new field that might get introduced

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py (hunk 25 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: :+1:

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py (hunk 26 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: :+1:

File: aria/modeling/service_template.py (hunk 27 above)
Comment (tliron) [inline aria/modeling/service_template.py]: :+1:

File: aria/modeling/service_template.py (hunk 28 above)
Comment (tliron) [inline aria/modeling/service_template.py]: You are correct, I had a long conversation with Maxim about it. Eventually when we have an instantiation module, it is very likely that the module will have a context. See comment at line 1861.

File: aria/orchestrator/workflows/api/task.py (hunk 30 above)
Comment (tliron) [inline aria/orchestrator/workflows/api/task.py]: What is the meaning of an empty implementation operation? The operation would fail in the orchestrator, seems much better to catch that here and not create that operation.

File: aria/orchestrator/execution_plugin/instantiation.py (hunk 31 above)
Comment (tliron) [inline aria/orchestrator/execution_plugin/instantiation.py]: Oops, leftover from devs. :+1:

File: aria/orchestrator/execution_plugin/instantiation.py (hunk 32 above)
Comment (tliron) [inline aria/orchestrator/execution_plugin/instantiation.py]: Well, I read our code and figured these out. :) They were not well documented anywhere.

File: aria/orchestrator/execution_plugin/instantiation.py (hunk 33 above)
Comment (tliron) [inline aria/orchestrator/execution_plugin/instantiation.py]: Due to the various validations. The structure is also a little bit different. In the service template, the user doesn't have to write "fabric_env" (or really shouldn't care about our implementation detail at all), the just write "ssh.password".

File: aria/orchestrator/execution_plugin/instantiation.py (hunk 35 above)
Comment (tliron) [inline aria/orchestrator/execution_plugin/instantiation.py]: I thought it would be more straightforward to do this than have a generic mechanism. Since *only* this plugin does this (for all the rest the configuration params get merged into inputs) it seemed unnecessary for me to create a generic mechanism.

File: aria/modeling/service_template.py (hunk 28 above)
Comment (ran-z) [inline aria/modeling/service_template.py]: I understand and I have seen the comment, but I'm saying the changes I've mentioned need to happen now, not after we have the instantiation module.
More so, I don't understand how this is expected to work anyway, as no execution context is expect to be present at the time this method is called.

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py [hunk 36]
```diff
@@ -677,21 +698,47 @@ def pattern(node_type, container): # pylint: disable=unused-argument
...
+        set_nested(the_dict[key], keys, value)
+
+
+def parse_implementation_string(context, service_template, presentation, model, implementation):
+    plugin_name, model.implementation = split_prefix(implementation)
+    if plugin_name is not None:
+        model.plugin_specification = service_template.plugin_specifications.get(plugin_name)
+        if model.plugin_specification is None:
+            context.validation.report(
+                'no policy for plugin "%s" specified in operation implementation: %s'
```
Comment (ran-z) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: use string format please :sweat_smile:

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py (hunk 23 above)
Comment (ran-z) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: Don't you think it'd be better to have it be set to `source` by default?
I assume when it's `None` we should treat it as `source`, so I think it'd be better if the data model reflects this design decision, no?

File: aria/orchestrator/workflows/builtin/utils.py (hunk 22 above)
Comment (ran-z) [inline aria/orchestrator/workflows/builtin/utils.py]: yes i know, I just figured this is not what you wanted to have there :sweat_smile: anyway i see its fixed, thanks

File: tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml (hunk 20 above)
Comment (ran-z) [inline tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml]: :+1:

File: aria/modeling/service_instance.py (hunk 18 above)
Comment (ran-z) [inline aria/modeling/service_instance.py]: oh, right. Please add a tiny comment next to it explaining that this case is only relevant for "Workflow operations" because I'm afraid I'll wonder about it the next time I see it as well :sweat_smile:  thanks!

File: aria/modeling/service_instance.py (hunk 12 above)
Comment (ran-z) [inline aria/modeling/service_instance.py]: :+1:

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py (hunk 36 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: :+1:

File: aria/orchestrator/execution_plugin/instantiation.py (hunk 32 above)
Comment (ran-z) [inline aria/orchestrator/execution_plugin/instantiation.py]: :+1:

File: aria/orchestrator/execution_plugin/instantiation.py (hunk 8 above)
Comment (ran-z) [inline aria/orchestrator/execution_plugin/instantiation.py]: :+1:

File: aria/modeling/service_instance.py (hunk 17 above)
Comment (ran-z) [inline aria/modeling/service_instance.py]: :+1:

File: aria/modeling/service_instance.py (hunk 16 above)
Comment (tliron) [inline aria/modeling/service_instance.py]: :+1:

File: aria/modeling/service_instance.py (hunk 18 above)
Comment (tliron) [inline aria/modeling/service_instance.py]: :+1:

File: tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml (hunk 20 above)
Comment (tliron) [inline tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml]: I'm changing this back. :) Will explain the new behavior on Slack.

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py (hunk 23 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: We agreed that it's better to keep this "edge was not set" information.

File: aria/orchestrator/workflows/api/task.py (hunk 30 above)
Comment (tliron) [inline aria/orchestrator/workflows/api/task.py]: We agreed to keep this behavior for now.

File: aria/modeling/service_template.py [hunk 37]
```diff
@@ -1864,13 +1865,21 @@ def as_raw(self):
 
     def instantiate(self, container):
         from . import models
-        plugin = self.plugin_specification.find_plugin() \
-            if self.plugin_specification is not None else None
+        if self.plugin_specification and self.plugin_specification.enabled:
+            plugin = self.plugin_specification.plugin
+            implementation = self.implementation if plugin is not None else None
+            # "plugin" would be none if a match was not found. In that case, a validation error
```
Comment (ran-z) [inline aria/modeling/service_template.py]: im not sure i understood this comment.
i think you're trying to say that `plugin` can't be None here since the test for enabled has already been performed earlier?
the first line of the comment makes it sound like `plugin` can be None when it cant

File: aria/modeling/service_template.py [hunk 38]
```diff
@@ -2077,27 +2091,39 @@ def service_template_fk(cls):
...
             ('name', self.name),
-            ('version', self.version)))
+            ('version', self.version),
+            ('enabled', self.enabled)))
 
     def coerce_values(self, container, report_issues):
         pass
 
     def instantiate(self, container):
-        from . import models
```
Comment (ran-z) [inline aria/modeling/service_template.py]: wait, so instantiate is now empty? :confused:
